### PR TITLE
DS-3406: Sort communities and collections iteration 2

### PR DESCRIPTION
**Pull request by tomdesair**
This pull request is an alternative fix for https://jira.duraspace.org/browse/DS-3406 which prevents Hibernate from executing `delete` and `insert` statements when sorting Collections or Communities.

### CODE REVIEW

File: dspace-api/src/test/java/org/dspace/content/CollectionTest.java
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/content/CollectionTest.java]: This is an odd place for a `commit()` (between a collection deletion & a community deletion). Could we add a comment here as to why this is needed? Is it because the Collection needs to be fully deleted prior to the Community? Or is this a mistake?

**Comment by tdonohue** [comment]
NOTE: This is also a fix for https://jira.duraspace.org/browse/DS-3599.

**Comment by tomdesair** [comment]
Good catch, the `commit()` is indeed unnecessary after I've removed the `@Sort` annotation. 

The problem is that a collection name can change after it was added to the set (especially in unit and integration tests) and this breaks the Set semantics, e.g.:

```
Community community = communityService.createSubcommunity(context, parent);
communityService.addMetadata(context, community, "dc", "title", null, null, "Restricted Community");
if(!parent.subCommunities.contains(community)) {
   System.out.println("Error").
}
```

This would always print "Error".

So I'm back to sorting the collections the moment you fetch them, but in a separate copy of the actual list so that Hibernate doesn't get confused.

**Comment by tdonohue** [comment]
@tomdesair : The code here looks good overall, though I'd like to see some enhanced Unit Tests to prove out the new logic.  For instance, I notice the tests for `Collection.getCommunities()`, `Community.getCollections()` and `Item.getCollections()` (and similar methods) all only test for the existence of one object. This may be a good opportunity to enhance those tests and ensure when multiple objects exist they are all returned, and in the expected order.

**Comment by pnbecker** [comment]
I reviewed and tested it. Code looks good, everything works as expected and I don't see any unexpected insert or delete sql commands.

I like the idea of @tdonohue to have more unit tests. Beside that: for me this PR looks ready to be merged. +1

**Comment by tdonohue** [comment]
@tomdesair : any chance you'll have time in the near future to try and enhance the unit tests here?  I believe that's all this PR is waiting on, and then we can merge it.

**Review by tdonohue — APPROVED** [review]
This has received two +1 votes and a test from Committers. Was just waiting on Unit Tests which have now been added. Merging this immediately

**Comment by tdonohue** [comment]
Sidenote: I did *not* cherrypick this to `master` as the original work (DS-3406) was also not cherry-picked to master. These both have now been noted as part of https://jira.duraspace.org/browse/DS-3580

**Comment by tdonohue** [comment]
UPDATE: this was cherry-picked to `master` in #1768

**Pull request merged**
